src/sidebar/util/view-filter.js

```diff
@@ -0,0 +1,195 @@
+/**
+ * Filter annotations against parsed search queries.
+ *
+ * When the user enters a query in the search box, the query is parsed using
+ * `generateFacetedFilter` and the currently loaded annotations are then matched
+ * against the parsed query by the `filter` method of this class. Annotations
+ * which do not match the filter are then hidden.
+ */
+
+/**
+ * @typedef {import('../../types/api').Annotation} Annotation
+ */
+
+import { quote } from './annotation-metadata';
+import * as unicodeUtils from './unicode';
+
+/**
+ * @typedef Filter
+ * @prop {(ann: Annotation) => boolean} matches
+ */
+
+/**
+ * @typedef Checker
+ * @prop {(ann: Annotation) => boolean} autofalse
+ * @prop {(ann: Annotation) => any|any[]} value
+ * @prop {(term: any, value: any) => boolean} match
+ */
+
+function displayName(ann) {
+  if (!ann.user_info) {
+    return '';
+  }
+  return ann.user_info.display_name || '';
+}
+
+/**
+ * Normalize a field value or query term for comparison.
+ */
+function normalize(val) {
+  if (typeof val !== 'string') {
+    return val;
+  }
+  return unicodeUtils.fold(unicodeUtils.normalize(val)).toLowerCase();
+}
+
+/**
+ * Filter that matches annotations against a single field & term.
+ *
+ * eg. "quote:foo" or "text:bar"
+ *
+ * @implements {Filter}
+ */
+class TermFilter {
+  /**
+   * @param {string} field - Name of field to match
+   * @param {string} term - Query term
+   * @param {Checker} checker - Functions for extracting term values from
+   *   an annotation and checking whether they match a query term.
+   */
+  constructor(field, term, checker) {
+    this.field = field;
+    this.term = term;
+    this.checker = checker;
+  }
+
+  matches(ann) {
+    const checker = this.checker;
+    if (checker.autofalse && checker.autofalse(ann)) {
+      return false;
+    }
+
+    let value = checker.value(ann);
+    if (Array.isArray(value)) {
+      value = value.map(normalize);
+    } else {
+      value = normalize(value);
+    }
+    return checker.match(this.term, value);
+  }
+}
+
+/**
+ * Filter that combines other filters using AND or OR combinators.
+ *
+ * @implements {Filter}
+ */
+class BinaryOpFilter {
+  /**
+   * @param {'and'|'or'} op - Binary operator
+   * @param {Filter[]} filters - Array of filters to test against
+   */
+  constructor(op, filters) {
+    this.operator = op;
+    this.filters = filters;
+  }
+
+  matches(ann) {
+    if (this.operator === 'and') {
+      return this.filters.every(filter => filter.matches(ann));
+    } else {
+      return this.filters.some(filter => filter.matches(ann));
+    }
+  }
+}
+
+/**
+ * Filters a set of annotations.
+ *
+ * @param {Annotation[]} annotations
+ * @param {Object} filters - Faceted filter generated by
+ * `generateFacetedFilter`.
+ * @return {string[]} IDs of matching annotations.
+ */
+export default function viewFilter(annotations, filters) {
```

Comment:
I'd suggest to rename this function so that it starts with a verb and name with a more obvious purpose (`filterAnnotations`?). This would require renaming the files as well.

Reply:
Really, I should have just done that :). The same thing occurred to me, too, I just didn't follow through. Will change.

Reply:
I've renamed this to `filterAnnotations`